icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java

```diff
@@ -0,0 +1,737 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+
+package com.ibm.icu.impl.units;
+
+import com.ibm.icu.util.*;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+
+public class MeasureUnitImpl {
+
+    /**
+     * The full unit identifier.  Null if not computed.
+     */
+    private String identifier = null;
+
+    /**
+     * The complexity, either SINGLE, COMPOUND, or MIXED.
+     */
+    private MeasureUnit.Complexity complexity = MeasureUnit.Complexity.SINGLE;
+
+    /**
+     * The list of simple units. These may be summed or multiplied, based on the
+     * value of the complexity field.
+     * <p>
+     * The "dimensionless" unit (SingleUnitImpl default constructor) must not be
+     * added to this list.
+     * <p>
+     * The "dimensionless" <code>MeasureUnitImpl</code> has an empty <code>singleUnits</code>.
+     */
+    private ArrayList<SingleUnitImpl> singleUnits;
+
+    public MeasureUnitImpl() {
+        singleUnits = new ArrayList<>();
+    }
+
+    public MeasureUnitImpl(SingleUnitImpl singleUnit) {
+        this();
+        this.appendSingleUnit(singleUnit);
+    }
+
+    /**
+     * Parse a unit identifier into a MeasureUnitImpl.
+     *
+     * @param identifier The unit identifier string.
+     * @return A newly parsed object.
+     * @throws <code>IllegalArgumentException</code> in case of incorrect/non-parsed identifier.
+     */
+    public static MeasureUnitImpl forIdentifier(String identifier) {
+        return UnitsParser.parseForIdentifier(identifier);
+    }
+
+    /**
+     * Used for currency units.
+     */
+    public static MeasureUnitImpl forCurrencyCode(String currencyCode) {
+        MeasureUnitImpl result = new MeasureUnitImpl();
+        result.identifier = currencyCode;
+        return result;
+    }
+
+    public MeasureUnitImpl clone() {
+        MeasureUnitImpl result = new MeasureUnitImpl();
+        result.complexity = this.complexity;
+        result.identifier = this.identifier;
+        result.singleUnits = (ArrayList<SingleUnitImpl>) this.singleUnits.clone();
```

Comment:
Looks like javac is giving me a warning for this. TBD: do we have Continuous Integration checks that warn upon error?

    [javac] /home/hugovdm/git/icu/icu4j/main/classes/core/src/com/ibm/icu/impl/units/MeasureUnitImpl.java:75: warning: [unchecked] unchecked cast
    [javac]         result.singleUnits = (ArrayList<SingleUnitImpl>) this.singleUnits.clone();
    [javac]                                                                                ^
    [javac]   required: ArrayList<SingleUnitImpl>
    [javac]   found:    Object
    [javac] Note: Some input files use or override a deprecated API.
    [javac] Note: Recompile with -Xlint:deprecation for details.
    [javac] 1 warning

Reply:
"The Java.util.ArrayList.clone() method is used to create a **shallow copy** of the mentioned array list. It just creates a copy of the list."

For the sake of getting my PR working, I'm replacing this line with:

        for (SingleUnitImpl single : this.singleUnits) {
            result.singleUnits.add(single.clone());
        }

Reply:
Yoshito is fixing this with PR https://github.com/unicode-org/icu/pull/1381. He also mentions @Cloneable, etc. I conclude I should insist on https://github.com/icu-units/icu/issues/93 - David shares lots of valuable Java knowledge that we should learn from.